### PR TITLE
Feature/redux

### PR DESCRIPTION
**Pull request by ghost**
## Summary

* Install and configure redux
* Get mockBooks connecting with redux
* Configure Reactotron to debug the store

#### Pixel 3XL
image

#### iPhone 8
image

## Trello Card

[Trello Card](https://trello.com/c/9xLwQxST/25-n-91-redux-i-actions-reducers-store)

### CODE REVIEW

File: wbooks/App.test.tsx [hunk 1]
```diff
@@ -8,6 +8,16 @@ import renderer from 'react-test-renderer';
...
 
+const reactotron = {
+  configure: () => reactotron,
+  use: () => reactotron,
+  connect: () => reactotron,
+  createEnhancer: () => reactotron,
+  setAsyncStorageHandler: () => reactotron
+};
+
+jest.mock('reactotron-react-native', () => reactotron);
```
Comment (sfernandez11) [inline wbooks/App.test.tsx]: We don't use this for mocks. Instead, create a `reactotron-react-native` file inside `__mocks_` folder and then
```
export default {
  configure: () => reactotron,
  use: () => reactotron,
  connect: () => reactotron,
  createEnhancer: () => reactotron,
  setAsyncStorageHandler: () => reactotron
}
```
But, is it really necessary?

File: wbooks/src/app/index.tsx [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
 import AppNavigator from '@components/AppNavigator';
+import store from '@redux/store';
 
 const App = () => {
   return (
-    <NavigationContainer>
-      <StatusBar barStyle="dark-content" />
-      <AppNavigator />
-    </NavigationContainer>
+    <Provider store={store}>
```
Comment (sfernandez11) [inline wbooks/src/app/index.tsx]: Move this to `App.tsx`:
```
export default function index() {
  return (
    <Provider store={store}>
      <App />
    </Provider>
  );
}
```

File: wbooks/src/app/screens/Library/index.tsx [hunk 3]
```diff
@@ -1,12 +1,20 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 import { SafeAreaView, FlatList, ListRenderItem } from 'react-native';
-import { Book } from '@interfaces/Book';
-import { BOOKS_MOCK } from '@constants/mockBooks';
+import { useDispatch, useSelector } from 'react-redux';
+import { Book, BookState } from '@interfaces/Book';
+import bookActions from '@redux/book/actions';
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: `bookActions` => `BookActions`

File: wbooks/src/app/screens/Library/index.tsx [hunk 4]
```diff
@@ -1,12 +1,20 @@
...
 
 import BookComponent from './components/Book';
 import styles from './styles';
 
 function Library() {
+  const dispatch = useDispatch();
+  const { books } = useSelector((state: BookState) => state);
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Use typed `useSelector`:
```
const books = useSelector<BookState, Book[]>(state => state.books);
```

File: wbooks/src/interfaces/Book.ts [hunk 5]
```diff
@@ -7,3 +7,8 @@ export interface Book {
   year: string;
   imageUrl?: string | null;
 }
+
+export interface BookState {
+  books: Book[];
+  book?: Book;
```
Comment (sfernandez11) [inline wbooks/src/interfaces/Book.ts]: Why this?

File: wbooks/src/redux/book/actions.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export const BOOK_ACTION_NAMES = {
+  getBooks: 'getBooks',
+  getBook: 'getBook'
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: Same, unnecessary

File: wbooks/src/redux/book/actions.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+};
+
+const BOOK_ACTION_CREATORS = {
+  getBooks: () => ({ type: BOOK_ACTION_NAMES.getBooks }),
+  getBook: (id: number) => ({ type: BOOK_ACTION_NAMES.getBook, payload: id })
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: Same

File: wbooks/src/redux/book/actions.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+export const BOOK_ACTION_NAMES = {
+  getBooks: 'getBooks',
+  getBook: 'getBook'
+};
+
+const BOOK_ACTION_CREATORS = {
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: `BOOK_ACTION_CREATORS` => `actionCreators`

File: wbooks/src/redux/book/actions.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+export const BOOK_ACTION_NAMES = {
+  getBooks: 'getBooks',
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: And the reducer identificator?

`getBooks: 'getBooks',` => `GET_BOOKS: '@@BOOK/GET_BOOKS'`

File: wbooks/src/redux/book/actions.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+export const BOOK_ACTION_NAMES = {
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: `BOOK_ACTION_NAMES` => `actions`

File: wbooks/src/redux/book/reducer.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import { BOOKS_MOCK } from '@constants/mockBooks';
+import { BookState } from '@interfaces/Book';
+
+import { BOOK_ACTION_NAMES } from './actions';
+
+const initialState: BookState = {
+  books: []
+};
+
+function bookReducer(state = initialState, action: any): BookState {
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts]: Why any? Fix TS

File: wbooks/src/redux/book/reducer.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  switch (action.type) {
+    case BOOK_ACTION_NAMES.getBooks:
+      return { ...state, books: BOOKS_MOCK };
+
+    case BOOK_ACTION_NAMES.getBook: {
+      const book = state.books.filter(obj => obj.id === action.payload)[0];
+      return { ...state, book };
+    }
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts]: Remove

File: wbooks/src/redux/book/reducer.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts]: Delete empty lines inside `bookReducer`

File: wbooks/src/redux/store.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
+import { createStore } from 'redux';
+import Reactotron from '@config/reactotronConfig';
+
+import bookReducer from './book/reducer';
+
+export default createStore(bookReducer, Reactotron.createEnhancer());
```
Comment (sfernandez11) [inline wbooks/src/redux/store.ts]: Add `Reactotron` only `if (__DEV__)`

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Some comments...

File: wbooks/App.test.tsx (hunk 1 above)
Comment (ghost) [reply wbooks/App.test.tsx]: If I do not mock reactotron, the test fail, but now I need to fix the test again, so I remove this mock for the moment

File: wbooks/src/redux/book/reducer.ts (hunk 11 above)
Comment (ghost) [reply wbooks/src/redux/book/reducer.ts]: Is this correct?
```
interface BookAction {
  type: string; 
} 
```

File: wbooks/src/redux/book/reducer.ts (hunk 12 above)
Comment (killermu94) [reply wbooks/src/redux/book/reducer.ts:18]: Why?, we do not need get a book detail?

File: wbooks/src/interfaces/Book.ts (hunk 5 above)
Comment (ghost) [reply wbooks/src/interfaces/Book.ts]: I need to know what's the correct type for book state

File: wbooks/src/app/index.tsx (hunk 2 above)
Comment (ghost) [reply wbooks/src/app/index.tsx]: Done!

File: wbooks/src/app/screens/Library/index.tsx (hunk 3 above)
Comment (ghost) [reply wbooks/src/app/screens/Library/index.tsx]: Done!

File: wbooks/src/app/screens/Library/index.tsx (hunk 4 above)
Comment (ghost) [reply wbooks/src/app/screens/Library/index.tsx]: Done!

File: wbooks/src/redux/book/actions.ts (hunk 6 above)
Comment (ghost) [reply wbooks/src/redux/book/actions.ts]: Done!

File: wbooks/src/redux/book/actions.ts (hunk 7 above)
Comment (ghost) [reply wbooks/src/redux/book/actions.ts]: Done!

File: wbooks/src/redux/book/actions.ts (hunk 8 above)
Comment (ghost) [reply wbooks/src/redux/book/actions.ts]: Done!

File: wbooks/src/redux/book/actions.ts (hunk 9 above)
Comment (ghost) [reply wbooks/src/redux/book/actions.ts]: Done!

File: wbooks/src/redux/book/actions.ts (hunk 10 above)
Comment (ghost) [reply wbooks/src/redux/book/actions.ts]: Done!

File: wbooks/src/redux/book/reducer.ts (hunk 12 above)
Comment (ghost) [reply wbooks/src/redux/book/reducer.ts]: Done!

File: wbooks/src/redux/book/reducer.ts (hunk 13 above)
Comment (ghost) [reply wbooks/src/redux/book/reducer.ts]: Done!

File: wbooks/src/redux/store.ts (hunk 14 above)
Comment (ghost) [reply wbooks/src/redux/store.ts]: Done!

File: wbooks/App.test.tsx (hunk 1 above)
Comment (ghost) [reply wbooks/App.test.tsx]: Done!

File: wbooks/src/redux/book/reducer.ts (hunk 11 above)
Comment (ghost) [reply wbooks/src/redux/book/reducer.ts]: Done!

**Review by sfernandez11 — APPROVED** [review]
🥇

**Pull request merged**